mlflow/gateway/providers/mosaicml.py

```diff
@@ -113,11 +114,14 @@ async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
         #    }
         #   }
         # }
-
-        resp = await self._request(
-            self.config.model.name,
-            final_payload,
-        )
+        try:
+            resp = await self._request(
+                self.config.model.name,
+                final_payload,
+            )
+        except HTTPException as e:
+            _custom_token_allowance_exceeded_handling(e.status_code, e.detail or {})
+            raise
```

Comment:
We can use a context manager here.

Reply:
/ai We want to capture `HTTPException` and change the status code to 422. `_custom_token_allowance_exceeded_handling` looks like this. Can you create a context manager for that so that we can reuse it?

```python
def _custom_token_allowance_exceeded_handling(status_code: int, detail):
    """
    Custom handler for specific error messages that are incorrectly set as server-side errors, but
    are in actuality an issue with the request sent to the external provider.
    """

    message = detail.get("message", {})
    if (
        status_code == 500
        and message
        and any(message.startswith(x) for x in MLFLOW_AI_GATEWAY_MOSAIC_ML_HANDLER_MESSAGES)
    ):
        raise HTTPException(status_code=422, detail=detail)
```

Reply:
/ai You nailed it

Reply:
I collapsed the handler and the context manager into a single handler function and simplified it, but it's pretty darned close to what I just wrote. Not bad, bot!